File: algorithms/arrays/LeetCode31/LeetCode31.java
package algorithms.arrays.LeetCode31;

/**
 * LeetCode Problem 31
 * Level: Medium
 * Title: Next Permutation
 * 
 * Description:
 * Implement next permutation, which rearranges numbers into the 
 * lexicographically next greater permutation of numbers. If such 
 * an arrangement is not possible, it must rearrange it as the 
 * lowest possible order (i.e., sorted in ascending order). The 
 * replacement must be in-place and use only constant extra memory.
 * 
 * Example:
 * Input: [1,2,3]
 * Output: [1,3,2]
 */
public class LeetCode31 {

    /**
     * Rearranges numbers into the lexicographically next greater permutation.
     * 
     * @param nums the input array of numbers
     */
    public void nextPermutation(int[] nums) {
        int n = nums.length;
        int i = n - 2;

        // Step 1 : Find first decreasing element
        while (i >= 0 && nums[i] >= nums[i + 1]) {
            i--;
        }

        // Step 2: Find element to swap with if step 1 is found
        if (i >= 0) {
            int j = n - 1;
            while (nums[j] <= nums[i]) {
                j--;
            }
            swap(nums, i, j);
        }

        // Step 3: Reverse the sequence from i + 1 to end
        reverse(nums, i + 1, n - 1);
    }

    /**
     * Swaps two elements in the array.
     * 
     * @param nums the array of numbers
     * @param i    index of the first element
     * @param j    index of the second element
     */
    private void swap(int[] nums, int i, int j) {
        int temp = nums[i];
        nums[i] = nums[j];
        nums[j] = temp;
    }

    /**
     * Reverses the elements in the array from start to end.
     * 
     * @param nums  the array of numbers
     * @param start the starting index
     * @param end   the ending index
     */
    private void reverse(int[] nums, int start, int end) {
        while (start < end) {
            swap(nums, start, end);
            start++;
            end--;
        }
    }
}

File: algorithms/arrays/leetcode31/LeetCode31.py
from typing import List

class Solution:
    def nextPermutation(self, nums: List[int]) -> None:
        """
        Do not return anything, modify nums in-place instead.
        """
        n = len(nums)
        if n <= 1:
            return
        
        # Step 1: Find the first index 'i' from the end such that nums[i] < nums[i+1]
        i = n - 2
        while i >= 0 and nums[i] >= nums[i + 1]:
            i -= 1
        
        # Step 2: If such index exists, find the first index 'j' from the end such that nums[j] > nums[i]
        if i >= 0:
            j = n - 1
            while nums[j] <= nums[i]:
                j -= 1
            # Swap nums[i] and nums[j]
            nums[i], nums[j] = nums[j], nums[i]
        
        # Step 3: Reverse the sub-array nums[i+1:]
        left, right = i + 1, n - 1
        while left < right:
            nums[left], nums[right] = nums[right], nums[left]
            left += 1
            right -= 1
